rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,108 @@
+- role
+  - edit
+- SSO
+  - connector removal
+  - user/group removal
+
+Those actions will cause a lock when:
+
+| action                 | initial state                                     | change                                                                                 |
+| ---------------------- | ------------------------------------------------- | -------------------------------------------------------------------------------------- |
+| user removal           | _admin_, the user with user deletion capabilities | user with user deletion capabilities deletes _admin_                                   |
+| user role removal      | _admin_                                           | _admin_ removes himself from a role that grants him user/role editing capabilities     |
+| role edit              | _admin_                                           | _admin_ edits own role to remove user/role editing capabilities                        |
+| SSO connector removal  | _SSO admin_                                       | _SSO admin_ removes SSO connector. _SSO admin_ logs out.                               |
+| SSO user/group removal | _SSO admin_                                       | Someone on the SSO side removes _SSO admin_ user or team/group containing _SSO admins_ |
+
+Teleport already protects against a similar case in the following scenario:
+
+| action       | initial state | change                           |
+| ------------ | ------------- | -------------------------------- |
+| role removal | _admin_       | _admin_ tries to remove own role |
+
+The error message is displayed: _failed to delete role that still in use by a user. Check system server logs for more details_.
+
+There is one more case when a user can lock himself out:
+
+| initial state | change                                                                                    |
+| ------------- | ----------------------------------------------------------------------------------------- |
+| _admin_       | _admin_ loses acess to own account (eg. he stores credentials and recovery codes locally) |
+
+However, requiring a second admin to guard this situation is not good for user experience. Existing and new clusters will have to add a second _admin_.
+
+## Terminology
+
+- _admin_: local user with user/role editing capabilities
+- _SSO admin_: user with user/role editing capabilities added via SSO connector (eg. GitHub)
+
+## Why
+
+When a user locks themselves out, they are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
+
+Further, for our Cloud offering, we don't perform account resets as a policy. The only recourse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data loss.
+
+On-prem customers could work around this "locked out" state by running `tctl` locally on the auth server to create a new admin user.
+
+## Details
```

Comment:
Some high-level feedback:

1. I think we need to discuss cluster types separately: on-prem local, on-prem SSO and Cloud. Of all of those, it seems that Cloud is the one that wants this feature the most - on-prem clusters can be accessed by other ways, meaning that `tctl` interventions are possible.

2. Given that Cloud is the one that wants this the most, let's talk about the "no reset" policy (and why we are still keeping it).

3. This is minor, but I find it better to have a design that states its preferred solution directly, moving alternatives to an "Alternatives Considered" section - that makes it clear what is the actual design vs what we considered (but didn't do).